src/main/java/com/github/manolo8/darkbot/core/utils/EntityList.java

```diff
@@ -8,229 +8,115 @@
-                } else {
-                    Entity entity = whenAdd(new Entity(id), found);
-                    if (entity.clickable.defRadius == 150) {
-                        npcs.add(whenAdd(new MapNpc(id), found)); // Experiment zone beacons
-                    } else {
-                        unknown.add(entity);
-                    }
-                }
-            }
-        }
-
-        /*if (main.config.BOT_SETTINGS.DEV_STUFF) {
-            long[] addr = Arrays.copyOf(entitiesAddress.elements, entitiesAddress.size);
-            Arrays.sort(addr);
-
-            for (int i = 0; i < addr.length; i++) {
-                long curr = addr[i];
-                if (!newFound.contains(curr)) continue;
-                int id = API.readMemoryInt(curr + 56);
-                if (id < 150_000_000 || id >= 150_200_000) continue;
-
-                StringBuilder str = new StringBuilder().append(id);
-                int search = 200;
-                if (i < addr.length - 1) search = Math.min(search, Math.max(100, (int) (addr[i+1] - curr)));
-                for (int offset = 0; offset < search; offset += 4) {
-                    int f = API.readMemoryInt(curr + offset);
-                    if (f > -1000 && f < 1000) str.append(" | ").append(offset).append(":").append(f);
-                }
-                System.out.println(str);
-            }
-        }*/
+        for (EntityFactory type : types)
+            this.entityListener.add(type, e -> list.add((T) e));
 
+        return list;
     }
 
-    private <E extends Entity> E whenAdd(E entity, long address) {
-
-        entity.added(main);
-        entity.update(address);
-        entity.update();
-
-        if (entity instanceof Obstacle)
-            obstacles.add((Obstacle) entity);
-
-        if (main.isRunning())
-            entity.clickable.setRadius(0);
-
-        return entity;
-    }
+    private void refreshEntities() {
+        entitiesArr.update();
+        for (int i = 0; i < entitiesArr.getSize(); i++) {
+            long entityPtr = entitiesArr.get(i);
 
-    private void whenRemove(Entity entity) {
-        entity.removed();
+            int id = API.readMemoryInt(entityPtr + 56);
+            if (ids.add(id)) entityListener.sendEntity(id, entityPtr);
+        }
     }
 
     private void removeAllInvalidEntities() {
         main.hero.pet.removed = main.hero.pet.isInvalid(address);
 
-        for (List<? extends Entity> entities : allEntities) {
-            for (int i = 0; i < entities.size(); i++) {
-                Entity entity = entities.get(i);
+        this.allEntities.forEach(entities -> entities.removeIf(entity -> {
+            if (entity.isInvalid(address) || entity.address == main.hero.address || entity.address == main.hero.pet.address) {
+                ids.remove(entity.id);
+                entity.removed();
+                return true;
+            } else entity.update();
```

Comment:
```suggestion
        for (List<? extends Entity> entities : allEntities) {
            // Remove invalid entities and update valid ones
            entities.removeIf(entity -> {
                if (entity.isInvalid(address) || entity.address == main.hero.address || entity.address == main.hero.pet.address) {
                    ids.remove(entity.id);
                    entity.removed();
                    return true;
                } 
                entity.update();
                return false;
            });
        }
```
I think it's better not to nest 2 lambdas here for readability